Re-deploy failed after cancel a deploy task

## Report

Thank you for trying **conjure-up**! Before reporting a bug please make sure you've gone through this checklist:

- [ ] Is this problem already documented at https://docs.ubuntu.com/conjure-up/en/troubleshoot#common-spell-problems
- [x] Is conjure-up running inside a virtual machine?
- [x] Is this problem reproducible with `sudo snap refresh conjure-up --edge`?

## Please provide the output of the following commands

```
which juju
> /snap/bin/juju

juju version
> 2.4.3-bionic-amd64

which conjure-up
> /snap/bin/conjure-up

conjure-up --version
> conjure-up 2.6.1

which lxc
> /snap/bin/lxc

/snap/bin/lxc config show
> config:
>  core.https_address: '[::]'

/snap/bin/lxc version
> Client version: 3.5
> Server version: 3.5

cat /etc/lsb-release
> DISTRIB_ID=Ubuntu
> DISTRIB_RELEASE=18.04
> DISTRIB_CODENAME=bionic
> DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
```

Please attach tarball of **~/.cache/conjure-up**:

```
tar cvzf conjure-up.tar.gz ~/.cache/conjure-up
```
https://transfer.sh/O9PQG/conjure-up.tar.gz

## Sosreport

Please attach a sosreport:

```
sudo apt install sosreport
sosreport

> sosreport (version 3.5)
> 
> no valid plugins were enabled

```

The resulting output file can be attached to this issue.


## What Spell was Selected?

## What provider (aws, maas, localhost, etc)?
localhost

### MAAS Users
Which version of MAAS?
N/A 

## Commands ran

Please outline what commands were run to install and execute conjure-up:
> conjure-up kubernetes

## Additional Information

When I try to deploy kubernetes in my KVM VPS, I found that I selected wrong network card, and it is hanging up at "Bootstrapping juju controller", so I pressed Quit to exit this process and re-entered conjure to select correct network card and deploy kubernetes. But it cannot work normally.  
Some important logs I think:  
```
2018-09-22 23:09:36,718 [DEBUG] conjure-up/kubernetes-core - step.py:268 - Executing script: /home/rinne/.cache/conjure-up/kubernetes-core/steps/04_enable-cni/before-config
2018-09-22 23:09:36,749 [DEBUG] conjure-up/kubernetes-core - telemetry.py:17 - Showing screen: Configure Applications
2018-09-22 23:09:39,086 [DEBUG] conjure-up/kubernetes-core - telemetry.py:17 - Showing screen: Creating Model
2018-09-22 23:09:39,202 [DEBUG] conjure-up/kubernetes-core - events.py:52 - Setting Error at conjureup/events.py:149
2018-09-22 23:09:39,203 [ERROR] conjure-up/kubernetes-core - events.py:161 - Unhandled exception in <Task finished coro=<BaseBootstrapController.run() done, defined at /snap/conjure-up/1022/lib/python3.6/site-packages/conjureup/controllers/juju/bootstrap/common.py:15> exception=IndexError('list index out of range',)>
Traceback (most recent call last):
  File "/snap/conjure-up/1022/lib/python3.6/site-packages/conjureup/controllers/juju/bootstrap/common.py", line 19, in run
    await self.do_add_model()
  File "/snap/conjure-up/1022/lib/python3.6/site-packages/conjureup/controllers/juju/bootstrap/common.py", line 24, in do_add_model
    if await juju.model_available():
  File "/snap/conjure-up/1022/lib/python3.6/site-packages/conjureup/juju.py", line 154, in model_available
    await controller.connect(app.provider.controller)
  File "/snap/conjure-up/1022/lib/python3.6/site-packages/juju/controller.py", line 100, in connect
    await self._connector.connect_controller(controller_name, **kwargs)
  File "/snap/conjure-up/1022/lib/python3.6/site-packages/juju/client/connector.py", line 90, in connect_controller
    endpoint = controller['api-endpoints'][0]
IndexError: list index out of range
2018-09-22 23:09:39,236 [DEBUG] conjure-up/kubernetes-core - __init__.py:27 - Showing dialog for exception: list index out of range
2018-09-22 23:09:39,262 [DEBUG] conjure-up/kubernetes-core - events.py:52 - Awaiting Bootstrapped at conjureup/controllers/juju/bootstrap/gui.py:43
2018-09-22 23:12:54,725 [DEBUG] conjure-up/kubernetes-core - events.py:52 - Setting Shutdown at conjureup/events.py:145
2018-09-22 23:12:54,727 [DEBUG] conjure-up/kubernetes-core - events.py:52 - Received Shutdown at conjureup/events.py:176
2018-09-22 23:12:54,728 [INFO] conjure-up/kubernetes-core - events.py:180 - Shutting down
```

--- a/conjureup/ui/views/ControllerListView.py
+++ b/conjureup/ui/views/ControllerListView.py
@@ -40,16 +40,19 @@ class ControllerListView(BaseView):
             widget.append(Padding.line_break(""))
             cdict = defaultdict(lambda: defaultdict(list))
             for cname, d in self.controllers.items():
-                cdict[d['cloud']][d.get('region', None)].append(cname)
+                cdict[d['cloud']][d.get('region', None)].append((cname, d))
 
             for cloudname, cloud_d in sorted(cdict.items()):
                 widget.append(Color.label(Text("  {}".format(cloudname))))
                 for regionname, controllers in cloud_d.items():
-                    for controller in sorted(controllers):
-                        label = "    {}".format(controller)
+                    for controller_name, controller in sorted(controllers):
+                        label = "    {}".format(controller_name)
                         if regionname:
                             label += " ({})".format(regionname)
-                        widget.append_option(label, controller)
+                        widget.append_option(
+                            label,
+                            controller,
+                            enabled=controller.get('api-endpoints'))
                 widget.append(Padding.line_break(""))
             widget.append(Padding.line_break(""))
         widget.append(HR())
@@ -57,5 +60,19 @@ class ControllerListView(BaseView):
         widget.select_first()
         return widget
 
+    def after_keypress(self):
+        selected = self.widget.selected_widgets
+        if selected is None:
+            return
+        elif selected.enabled:
+            msg = self.footer
+        else:
+            msg = ('This controller has no endpoints, so it cannot be used. '
+                   'To clean it up, run: juju unregister {}'.format(
+                       selected.label.strip().split()[0]))
+        self.set_footer(msg)
+
     def submit(self):
+        if not self.widget.selected:
+            return  # tried to select disabled controller
         self.submit_cb(self.widget.selected)

--- a/conjureup/utils.py
+++ b/conjureup/utils.py
@@ -612,7 +612,7 @@ def is_valid_hostname(hostname):
     if hostname[-1] == ".":
         # strip exactly one dot from the right, if present
         hostname = hostname[:-1]
-    allowed = re.compile("(?!-)[A-Z\d\-\_]{1,63}(?<!-)$", re.IGNORECASE)
+    allowed = re.compile(r"(?!-)[A-Z\d\-\_]{1,63}(?<!-)$", re.IGNORECASE)
     return all(allowed.match(x) for x in hostname.split("."))
 
 

--- a/tools/graph-controllers.py
+++ b/tools/graph-controllers.py
@@ -5,8 +5,9 @@ import os
 import re
 from subprocess import PIPE, run
 
-controllers_use_re = re.compile("(\s*)\S*\s*controllers\.use\('(.*)'\).render")
-scope_re = re.compile("^(\s*)def (\w*)\(")
+controllers_use_re = re.compile(
+    r"(\s*)\S*\s*controllers\.use\('(.*)'\).render")
+scope_re = re.compile(r"^(\s*)def (\w*)\(")
 
 default_name_map = dict(spellpicker="Spell Selection",
                         bundlereadme="Spell Readme",

--- a/tox.ini
+++ b/tox.ini
@@ -21,7 +21,7 @@ commands =
     pylint conjureup test tools
 
 [testenv:flake]
-commands = flake8 --ignore E501,E722,E741 {posargs} conjureup test tools
+commands = flake8 --ignore E501,E722,E741,W504 {posargs} conjureup test tools
 deps = flake8
 
 [testenv:docs]

